src/data/nat/up.lean

```diff
@@ -0,0 +1,74 @@
+/-
+Copyright (c) 2020 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Simon Hudon
+-/
+import data.nat.basic
+
+/-!
+# `nat.up`
+
+`nat.up p`, with `p` a predicate on `ℕ`, is a subtype of `ℕ` that contains value
+`n` if no value below `n` (excluding `n`) satisfies `p`.
+
+This allows us to prove `>` is well-founded when `∃ i, p i`. This helps implement
+searches on `ℕ`, starting at `0` and with an unknown upper-bound.
```

Comment:
How does this differ from `nat.find`? Why do we need another definition for this?

Reply:
`nat.find` requires a decidable predicate.

Reply:
So does `nat.up.find`...

Reply:
Right, I don't actually need a new definition of `find`. I only need `nat.up` with its well-founded order.

Reply:
Could you mention the similarities/differences in the module doc?